### PR TITLE
ocaml-freestanding becomes a cross-compiler: take 2

### PR DESCRIPTION
**Pull request by TheLortex**
This PR replaces https://github.com/mirage/ocaml-freestanding/pull/90. The introduction of the PR is still relevant.

Thanks to @mato's work on solo5 (https://github.com/Solo5/solo5/pull/494), we finally have a good and future-proof scheme for ocaml-freestanding. What happens is that for each solo5 _toolchain_, we will install an ocaml compiler distribution configured with that toolchain. For now, we only have one toolchain which is `solo5-none-static`. 

From the toolchain name is derived a compiler and a linker to use: `x86_64-solo5-none-static-cc` and `x86_64-solo5-none-static-ld` which are used to build a freestanding libc/libm and a freestanding cross-compiler. The distribution is installed under the ocamlfind toolchain name `freestanding`.

This cross-compiler can then be used by ocaml projects to build solo5 unikernels. I've added an `example/` folder to display the layout of such a project. 
**Warning:** existing projects depending on ocaml-freestanding will be **broken**, because they will use the host compiler instead of `ARCH-solo5-none-static-cc`. These projects should be updated so that they use `pkg-config --var=cc ocaml-freestanding` to figure out the correct compiler to use. 

Because of the lack of unix functions, `libcamlrun.a` is not built. So only native code compilation is supported. I've added CI to test it on `4.10`, `4.11` and `4.12` ubuntu, and updated the cirrus CI for freebsd testing.

### CODE REVIEW

File: nolibc/sysdeps_solo5.c [hunk 1]
```diff
@@ -119,6 +119,12 @@ void *sbrk(intptr_t increment)
     return (void *)prev;
 }
 
+int __getauxval(int unused) {
+    errno = ENOENT;
+    return 0;
+}
+
+
```
Comment (dinosaure) [inline nolibc/sysdeps_solo5.c:127]: May be put a comment about this function to explain why we add it.

Comment (TheLortex) [reply nolibc/sysdeps_solo5.c:133]: Thanks, I added a commit to document that.

**Comment by TheLortex** [comment]
I've added a new commit for the mirage 3 compatibility. Currently freestanding C stubs use `pkg-config --cflags ocaml-freestanding` to find the correct flags, but the new solo5 changes break this behavior as the C stubs should be built using the solo5 compiler toolchain. 
So we have a choice to make:
- modify all packages that currently build freestanding C stubs to use the solo5 compiler toolchain instead of the regular cc (`bigstringaf`, `mirage-crypto`, `tcpip`, ...) and cut patch releases to keep the mirage 3 compat
- modify the flags exposed by `pkg-config --cflags ocaml-freestanding` to bring back the old behavior. This is what this additional commit does. It relies on solo5 installing a `.pc` file to describe its toolchain flags. This way, from a mirage 3 point of view, the transition to the new ocaml-freestanding is transparent.

**Comment by dinosaure** [comment]
> modify the flags exposed by pkg-config --cflags ocaml-freestanding to bring back the old behavior. This is what this additional commit does. It relies on solo5 installing a .pc file to describe its toolchain flags. This way, from a mirage 3 point of view, the transition to the new ocaml-freestanding is transparent.

Thanks for this update, as I said, both options are fine in the scope of the release of MirageOS 4 - even if the second one is better (because it's transparent). I hope, it's the last contentious piece between MirageOS 3 & MirageOS 4.

**Comment by dinosaure** [comment]
Rebased with `master` :+1:.

File: Makefile [hunk 2]
```diff
@@ -1,24 +1,55 @@
...
+	    "SYSDEP_OBJS=$(MAKECONF_NOLIBC_SYSDEP_OBJS)"
 
+# OPENLIBM
 openlibm/libopenlibm.a:
-	$(MAKE) -C openlibm "CFLAGS=$(FREESTANDING_CFLAGS)" libopenlibm.a
+	$(MAKE) -C openlibm "CC=$(MAKECONF_CC)" "CPPFLAGS=$(LOCAL_CFLAGS)" libopenlibm.a
 
+# OCAML
 ocaml/Makefile:
 	cp -r `ocamlfind query ocaml-src` ./ocaml
```
Comment (hannesm) [inline Makefile:30]:
```suggestion
	cp -r `ocamlfind query ocaml-src` ./ocaml
#configure avoid inserting -lm
	sed -i -e 's/ac_cv_lib_m_cos=yes/ac_cv_lib_m_cos=no/' ocaml/configure
```

as learned by #101 // #102

File: Makefile [hunk 3]
```diff
@@ -32,60 +63,80 @@ ocaml/Makefile:
...
-# - HAS_XXX must be defined manually since our invocation of configure cannot
-#   link against nolibc (which would need to produce complete Solo5 binaries).
+# - We use LIBS with a stubbed out solo5 implementation to override the OCaml 
+# 	configure link test
 # - We override OCAML_OS_TYPE since configure just hardcodes it to "Unix".
-OCAML_CFLAGS=$(FREESTANDING_CFLAGS) -I$(TOP)/openlibm/include -I$(TOP)/openlibm/src
-
-ocaml/Makefile.config: ocaml/Makefile
+# - We override HAS_SOCKETS because of a bug in the ocaml configure script that
+# 	always enables sockets.
```
Comment (hannesm) [inline Makefile:70]: maybe mention "https://github.com/ocaml/ocaml/pull/10156" and "can be removed, once support for OCaml < 4.12.1 is dropped" (or directly drop that support?)

File: Makefile (hunk 2 above)
Comment (hannesm) [reply Makefile:30]: alternatively, we can pass `ac_cv_lib_m_cos=no` to configure below (not tested)

File: Makefile [hunk 4]
```diff
@@ -32,60 +63,80 @@ ocaml/Makefile:
...
+		$(MAKECONF_OCAML_CONFIGURE_OPTIONS)
+	echo "ARCH=$(MAKECONF_OCAML_BUILD_ARCH)" >> ocaml/Makefile.config
 	echo 'SAK_CC=cc' >> ocaml/Makefile.config
-	echo 'SAK_CFLAGS=$(OC_CFLAGS) $(OC_CPPFLAGS)' >> ocaml/Makefile.config
-	echo 'SAK_LINK=$(SAK_CC) $(SAK_CFLAGS) $(OUTPUTEXE)$(1) $(2)' >> ocaml/Makefile.config
-	echo '#define HAS_GETTIMEOFDAY' >> ocaml/runtime/caml/s.h
-	echo '#define HAS_SECURE_GETENV' >> ocaml/runtime/caml/s.h
-	echo '#define HAS_TIMES' >> ocaml/runtime/caml/s.h
+	echo 'SAK_CFLAGS=' >> ocaml/Makefile.config
+	echo 'SAK_LINK=cc $(SAK_CFLAGS) $$(OUTPUTEXE)$$(1) $$(2)' >> ocaml/Makefile.config
```
Comment (hannesm) [inline Makefile:96]: I'm not sure why the SAK_ stuff changed here?

File: Makefile [hunk 5]
```diff
@@ -32,60 +63,80 @@ ocaml/Makefile:
...
+	echo '#undef HAS_SOCKETS' >> ocaml/runtime/caml/s.h
```
Comment (hannesm) [inline Makefile:97]:
```suggestion
```
and lower bound to ocaml 4.12.1

File: Makefile [hunk 6]
```diff
@@ -32,60 +63,80 @@ ocaml/Makefile:
...
 	echo '#define OCAML_OS_TYPE "None"' >> ocaml/runtime/caml/s.h
 
 ocaml/runtime/caml/version.h: ocaml/Makefile.config
 	ocaml/tools/make-version-header.sh > $@
 
-ocaml/runtime/libasmrun.a: ocaml/Makefile.config ocaml/runtime/caml/version.h
-	$(MAKE) -C ocaml/runtime libasmrun.a
+CAMLOPT:=$(shell which ocamlopt)
+CAMLRUN:=$(shell which ocamlrun)
+CAMLC:=$(shell which ocamlc)
```
Comment (hannesm) [inline Makefile:106]: these 3 definitions are not used anywhere, or am I wrong?

File: Makefile [hunk 7]
```diff
@@ -32,60 +63,80 @@ ocaml/Makefile:
...
+# CONFIGURATION FILES
 ocaml-freestanding.pc: ocaml-freestanding.pc.in Makeconf
-	sed -e 's!@@PKG_CONFIG_DEPS@@!$(PKG_CONFIG_DEPS)!' \
-	    -e 's!@@PKG_CONFIG_EXTRA_LIBS@@!$(PKG_CONFIG_EXTRA_LIBS)!' \
+	sed -e 's!@@PKG_CONFIG_EXTRA_LIBS@@!$(MAKECONF_PKG_CONFIG_EXTRA_LIBS)!' \
+	    -e 's!@@PKG_CONFIG_CC@@!$(MAKECONF_CC)!' \
+	    -e 's!@@PKG_CONFIG_LD@@!$(MAKECONF_LD)!' \
+		-e 's!@@PKG_CONFIG_SOLO5_TOOLCHAIN@@!$(MAKECONF_TOOLCHAIN)!' \
+	    -e 's!@@CFLAGS@@!$(MAKECONF_CFLAGS)!' \
 	    ocaml-freestanding.pc.in > $@
```
Comment (hannesm) [inline Makefile:119]: since we decided to drop pkg-config support, the rules above can be removed.

File: Makefile [hunk 8]
```diff
@@ -32,60 +63,80 @@ ocaml/Makefile:
...
-	./install.sh
+	MAKE=$(MAKE) PREFIX=$(MAKECONF_PREFIX) ./install.sh
 
 uninstall:
 	./uninstall.sh
 
 clean:
-	-$(MAKE) -C ocaml/runtime clean
+	$(RM) -r ocaml/
+	$(RM) ocaml-freestanding.pc freestanding.conf
```
Comment (hannesm) [inline Makefile:134]:
```suggestion
	$(RM) freestanding.conf
```

File: README.md [hunk 9]
```diff
@@ -23,27 +23,69 @@ on your system as a dependency when you build this package.
...
+- `libnolibc.a`: libc interfaces required by the OCaml runtime.
+- `libopenlibm.a`: libm required by the OCaml runtime.
 
-In `PREFIX/include/ocaml-freestanding/include`:
+In `PREFIX/frestanding-sysroot/include/nolibc`:
 
 - Header files for nolibc and openlibm.
 
-In `PREFIX/include/ocaml-freestanding/include/caml`:
+In `PREFIX/lib/pkgconfig`:
```
Comment (hannesm) [inline README.md:47]:
```suggestion
```

File: README.md [hunk 10]
```diff
@@ -23,27 +23,69 @@ on your system as a dependency when you build this package.
...
+
```
Comment (hannesm) [inline README.md:48]:
```suggestion
```

File: README.md [hunk 11]
```diff
@@ -23,27 +23,69 @@ on your system as a dependency when you build this package.
...
+- `ocaml-freestanding.pc`: package definition to link the ocaml runtime.
```
Comment (hannesm) [inline README.md:49]:
```suggestion
```

File: README.md [hunk 12]
```diff
@@ -23,27 +23,69 @@ on your system as a dependency when you build this package.
...
+
```
Comment (hannesm) [inline README.md:50]:
```suggestion
```

File: README.md [hunk 13]
```diff
@@ -23,27 +23,69 @@ on your system as a dependency when you build this package.
...
+The `example` describes the minimal structure needed to build an 
+ocaml-freestanding executable with dune, linked with the hvt bindings. It 
+requires an application manifest and a startup file to initialize the libc. 
+
+Build: `dune build -x freestanding` 
+Run: `solo5-hvt _build/default.freestanding/main.exe`
 
-- Header files for OCaml runtime.
+### Usage - the old way (< 0.7.0):
```
Comment (hannesm) [inline README.md:77]: this section should be removed.

File: README.md [hunk 14]
```diff
@@ -23,27 +23,69 @@ on your system as a dependency when you build this package.
...
+and linking C code:
+- CFLAGS: `pkg-config --cflags ocaml-freestanding`
+- CC: `pkg-config --var=ld ocaml-freestanding`
+- LIBS: `pkg-config --libs ocaml-freestanding`
+- LD: `pkg-config --var=cc ocaml-freestanding`
 
 ## Supported compiler versions
 
-Tested against OCaml 4.08.0 through 4.13.0. Other versions may require
+Tested against OCaml 4.10.0 through 4.13.0. Other versions may require
```
Comment (hannesm) [inline README.md:88]: update to 4.12.1

File: install.sh [hunk 15]
```diff
@@ -4,11 +4,13 @@ prefix=${1:-$PREFIX}
 if [ "$prefix" = "" ]; then
     prefix=`opam config var prefix`
```
Comment (hannesm) [inline install.sh:5]:
```suggestion
    prefix=`opam var prefix`
```

File: install.sh [hunk 16]
```diff
@@ -30,3 +30,13 @@ touch ${DESTLIB}/META
 # pkg-config
```
Comment (hannesm) [inline install.sh:30]:
```suggestion
```

File: install.sh [hunk 17]
```diff
@@ -30,3 +30,13 @@ touch ${DESTLIB}/META
...
 mkdir -p ${prefix}/lib/pkgconfig
```
Comment (hannesm) [inline install.sh:31]:
```suggestion
```

File: install.sh [hunk 18]
```diff
@@ -30,3 +30,13 @@ touch ${DESTLIB}/META
...
 cp ocaml-freestanding.pc ${prefix}/lib/pkgconfig/ocaml-freestanding.pc
```
Comment (hannesm) [inline install.sh:32]:
```suggestion
```

File: install.sh [hunk 19]
```diff
@@ -30,3 +30,13 @@ touch ${DESTLIB}/META
...
+
```
Comment (hannesm) [inline install.sh:33]:
```suggestion
```

File: ocaml-freestanding-cross-aarch64.opam [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+bug-reports: "https://github.com/mirage/ocaml-freestanding/issues/"
+license: "MIT"
+tags: "org:mirage"
+dev-repo: "git+https://github.com/mirage/ocaml-freestanding.git"
+build: [
+  ["./configure.sh"
+   "--prefix=%{prefix}%"
+   "--target=aarch64-solo5-none-static"
+   "--ocaml-configure-option=--with-afl" {ocaml-option-afl:installed}
+   "--ocaml-configure-option=--disable-native-compiler" {ocaml-option-bytecode-only:installed}
```
Comment (hannesm) [inline ocaml-freestanding-cross-aarch64.opam:14]: is this useful? any bytecode unikernels?

File: ocaml-freestanding-cross-aarch64.opam [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+   "--ocaml-configure-option=--disable-force-safe-string" {ocaml-option-default-unsafe-string:installed}
```
Comment (hannesm) [inline ocaml-freestanding-cross-aarch64.opam:15]: I'd remove this... safe-string is a very good thing to have.

File: ocaml-freestanding-cross-aarch64.opam [hunk 22]
```diff
@@ -0,0 +1,53 @@
...
+   "--ocaml-configure-option=DEFAULT_STRING=unsafe" {ocaml-option-default-unsafe-string:installed}
```
Comment (hannesm) [inline ocaml-freestanding-cross-aarch64.opam:16]: same as above

File: ocaml-freestanding-cross-aarch64.opam [hunk 23]
```diff
@@ -0,0 +1,53 @@
...
+   "--ocaml-configure-option=--disable-flat-float-array" {ocaml-option-no-flat-float-array:installed}
+   "--ocaml-configure-option=--enable-flambda" {ocaml-option-flambda:installed}
+   "--ocaml-configure-option=--enable-frame-pointers" {ocaml-option-fp:installed}
+   "--ocaml-configure-option=--enable-spacetime" {ocaml-option-spacetime:installed}
```
Comment (hannesm) [inline ocaml-freestanding-cross-aarch64.opam:20]: I suspect there's not much use of spacetime, plus it was removed from OCaml recently... maybe just drop this option as well?

File: ocaml-freestanding-cross-aarch64.opam [hunk 24]
```diff
@@ -0,0 +1,53 @@
...
+   "--ocaml-configure-option=--disable-naked-pointers" {ocaml-option-nnp:installed}
+   "--ocaml-configure-option=--enable-naked-pointers-checker" {ocaml-option-nnpchecker:installed}
+  ]
+  [make "-j%{jobs}%"]
+]
+install: [make "install" ]
+depends: [
+  "ocamlfind" {build}
+  "ocaml-src" {build}
+  "ocaml" {>= "4.08.0" & < "4.14.0"}
```
Comment (hannesm) [inline ocaml-freestanding-cross-aarch64.opam:30]:
```suggestion
  "ocaml" {>= "4.12.1" & < "4.14.0"}
```

File: ocaml-freestanding.pc.in [hunk 25]
```diff
@@ -1,12 +1,13 @@
 prefix=${pcfiledir}/../..
```
Comment (hannesm) [inline ocaml-freestanding.pc.in:1]: remove this file

File: test/dune [hunk 26]
```diff
@@ -0,0 +1,41 @@
+
```
Comment (hannesm) [inline test/dune:1]: leading empty lines can be safely removed

File: uninstall.sh [hunk 27]
```diff
@@ -7,5 +7,5 @@ fi
 
 odir=$prefix/lib
 rm -f $odir/pkgconfig/ocaml-freestanding.pc
```
Comment (hannesm) [inline uninstall.sh:9]:
```suggestion
```

File: configure.sh [hunk 28]
```diff
@@ -1,60 +1,110 @@
...
+        echo "ERROR: Unsupported architecture: ${BUILD_TRIPLET}" 1>&2
         exit 1
         ;;
 esac
 
-PKG_CONFIG_EXTRA_LIBS=
+EXTRA_LIBS=
 if [ "${BUILD_ARCH}" = "aarch64" ]; then
-    PKG_CONFIG_EXTRA_LIBS="$PKG_CONFIG_EXTRA_LIBS $(gcc -print-libgcc-file-name)" || exit 1
+    EXTRA_LIBS="$EXTRA_LIBS -lgcc" || exit 1
```
Comment (hannesm) [inline configure.sh:95]: I'm not sure about this change. but I assume you've tested this :)

**Comment by hannesm** [comment]
I addressed most comments in #104 - remaning are SAK changes and arm64 libgcc. feel free to merge #104 here or continue there. This all looks fine to be merged and released.

File: configure.sh (hunk 28 above)
Comment (dinosaure) [reply configure.sh:95]: So we discovered with @TheLortex that on aarch64, `gcc` emits some functions which used some internals only available by the _fake_ library `libgcc` (which does not exists).

It's a bit strange but we must link to it via `-lgcc`. We should however look and document more precisely that indeed.

File: Makefile (hunk 4 above)
Comment (dinosaure) [reply Makefile:96]: Hmmhmm that's true. Do you have any clues @TheLortex?

Comment (dinosaure) [reply Makefile:96]: Hmmhm, this update comes from my commit. If I remember correctly, `SAK_CC` is used to compile OCaml and used by default `CC` if this variable is defined. About `SAK_CFLAGS`, it's pretty the same, it uses `CFLAGS`/`OC_FLAGS`. These flags are correctly updated before and we should not _redo_ what the OCaml's `configure` does.

**Comment by dinosaure** [comment]
Close by #104

**Pull request closed without merging**